src/header/shared/entity.rs

```diff
@@ -1,41 +1,114 @@
 use std::str::FromStr;
 use std::fmt::{self, Display};
 
-/// An entity tag
+// check that each char in the slice is either:
+// 1. %x21, or
+// 2. in the range %x23 to %x7E, or
+// 3. in the range %x80 to %xFF
+fn check_slice_validity(slice: &str) -> bool {
+    for c in slice.bytes() {
+        match c {
+            b'\x21' | b'\x23' ... b'\x7e' | b'\x80' ... b'\xff' => (),
+            _ => { return false; }
+        }
+    }
+    true
+}
+
+/// An entity tag, defined in [RFC7232](https://tools.ietf.org/html/rfc7232#section-2.3)
 ///
-/// An Etag consists of a string enclosed by two literal double quotes.
+/// An entity tag consists of a string enclosed by two literal double quotes.
 /// Preceding the first double quote is an optional weakness indicator,
-/// which always looks like this: W/
-/// See also: https://tools.ietf.org/html/rfc7232#section-2.3
-#[derive(Clone, PartialEq, Debug)]
+/// which always looks like `W/`. Examples for valid tags are `"xyzzy"` and `W/"xyzzy"`.
+///
+/// # ABNF
+/// ```plain
+/// entity-tag = [ weak ] opaque-tag
+/// weak       = %x57.2F ; "W/", case-sensitive
+/// opaque-tag = DQUOTE *etagc DQUOTE
+/// etagc      = %x21 / %x23-7E / obs-text
+///            ; VCHAR except double quotes, plus obs-text
+/// ```
+///
+/// # Comparison
+/// To check if two entity tags are equivalent in an application always use the `strong_eq` or
+/// `weak_eq` methods based on the context of the Tag. Only use `==` to check if two tags are
+/// identical.
+///
+/// The example below shows the results for a set of entity-tag pairs and
+/// both the weak and strong comparison function results:
+///
+/// | ETag 1  | ETag 2  | Strong Comparison | Weak Comparison |
+/// |---------|---------|-------------------|-----------------|
+/// | `W/"1"` | `W/"1"` | no match          | match           |
+/// | `W/"1"` | `W/"2"` | no match          | no match        |
+/// | `W/"1"` | `"1"`   | no match          | match           |
+/// | `"1"`   | `"1"`   | match             | match           |
+#[derive(Clone, Debug, Eq, PartialEq)]
 pub struct EntityTag {
     /// Weakness indicator for the tag
     pub weak: bool,
     /// The opaque string in between the DQUOTEs
-    pub tag: String
+    tag: String
 }
 
-impl Display for EntityTag {
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        if self.weak {
-            try!(write!(fmt, "{}", "W/"));
+impl EntityTag {
+    /// Constructs a new EntityTag.
+    /// # Panics
+    /// If the tag contains invalid characters.
+    pub fn new(weak: bool, tagslice: &str) -> EntityTag {
```

Comment:
Ah, yes.

Specifically, the only way to pass a `&str` is to have a static str, or they need to create a `String`. Since they have already allocated a `String` in order pass a reference, it's wasteful to allocate a new String, especially since they likely just threw away the original they made.